Disable parallelize_output_from_storages for processing MATERIALIZED VIEWs and storages with one block only

azat opened the pull request:
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

### Documentation entry for user-facing changes
Disable `parallelize_output_from_storages` for processing MATERIALIZED VIEWs and storages with one block only

Adding more processors for parallelize_output_from_storages is not a
costless operation (I've experienced some issues in production because
of this), and it is not easy to fix in a normal way, so let's disable it
for now.

#### Example for MATERIALIZED VIEWs

```
INSERT INTO input SELECT * FROM numbers(10e6) SETTINGS parallelize_output_from_storages=0, min_insert_block_size_rows=1000
  0 rows in set. Elapsed: 1.851 sec. Processed 20.00 million rows, 120.00 MB (10.80 million rows/s., 64.82 MB/s.)

INSERT INTO input SELECT * FROM numbers(10e6) SETTINGS parallelize_output_from_storages=1, min_insert_block_size_rows=1000
  0 rows in set. Elapsed: 3.648 sec. Processed 20.00 million rows, 120.00 MB (5.48 million rows/s., 32.90 MB/s.)
```

#### Example for Values

```
clickhouse-benchmark -i 10000 -d 0 --parallelize_output_from_storages=0 -q "select * from values('foo')"
localhost:9000, queries 10000, QPS: 2800.490, RPS: 2800.490, MiB/s: 0.032, result RPS: 2800.490, result MiB/s: 0.032.

clickhouse-benchmark -i 10000 -d 0 --parallelize_output_from_storages=1 -q "select * from values('foo')"
localhost:9000, queries 10000, QPS: 1259.805, RPS: 1259.805, MiB/s: 0.014, result RPS: 1259.805, result MiB/s: 0.014.
```

Note: no need to backport, since the last PR from this group (#49479) is not included into any release, yet
Follow-up for: #49479 (cc @al13n321)
Follow-up for: #48727 (cc @devcrafter)

azat committed:
Disable parallelize_output_from_storages for processing MATERIALIZED VIEWs

Adding more processors for parallelize_output_from_storages is not a
costless operation (I've experienced some issues in production because
of this), and it is not easy to fix in a normal way, so let's disable it
for now.

Before this patch:
- INSERT INTO input SELECT * FROM numbers(10e6) SETTINGS parallelize_output_from_storages=1, min_insert_block_size_rows=1000
  0 rows in set. Elapsed: 3.648 sec. Processed 20.00 million rows, 120.00 MB (5.48 million rows/s., 32.90 MB/s.)

- INSERT INTO input SELECT * FROM numbers(10e6) SETTINGS parallelize_output_from_storages=0, min_insert_block_size_rows=1000
  0 rows in set. Elapsed: 1.851 sec. Processed 20.00 million rows, 120.00 MB (10.80 million rows/s., 64.82 MB/s.)

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Disable parallelize_output_from_storages for storages with only one block

    $ clickhouse-benchmark -i 10000 -d 0 --parallelize_output_from_storages=0 -q "select * from values('foo')"
    Loaded 1 queries.

    Queries executed: 10000.

    localhost:9000, queries 10000, QPS: 2800.490, RPS: 2800.490, MiB/s: 0.032, result RPS: 2800.490, result MiB/s: 0.032.

    0.000%          0.000 sec.
    10.000%         0.000 sec.
    20.000%         0.000 sec.
    30.000%         0.000 sec.
    40.000%         0.000 sec.
    50.000%         0.000 sec.
    60.000%         0.000 sec.
    70.000%         0.000 sec.
    80.000%         0.000 sec.
    90.000%         0.000 sec.
    95.000%         0.000 sec.
    99.000%         0.001 sec.
    99.900%         0.001 sec.
    99.990%         0.001 sec.

    $ clickhouse-benchmark -i 10000 -d 0 --parallelize_output_from_storages=1 -q "select * from values('foo')"
    Loaded 1 queries.

    Queries executed: 10000.

    localhost:9000, queries 10000, QPS: 1259.805, RPS: 1259.805, MiB/s: 0.014, result RPS: 1259.805, result MiB/s: 0.014.

    0.000%          0.001 sec.
    10.000%         0.001 sec.
    20.000%         0.001 sec.
    30.000%         0.001 sec.
    40.000%         0.001 sec.
    50.000%         0.001 sec.
    60.000%         0.001 sec.
    70.000%         0.001 sec.
    80.000%         0.001 sec.
    90.000%         0.001 sec.
    95.000%         0.001 sec.
    99.000%         0.001 sec.
    99.900%         0.001 sec.
    99.990%         0.003 sec.

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

robot-clickhouse-ci-2 commented:
<!-- automatic status comment for PR #50214 from azat/ClickHouse:parallelize_output_from_storages-fix -->
This is an automated comment for commit e287ec50920c7cadabea6ec19ef14b353345ac93 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/50214/e287ec50920c7cadabea6ec19ef14b353345ac93/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

azat commented:
Test failures looks unrelated:

>Stateless tests (release, analyzer) — fail: 2, passed: 5281, skipped: 19

- `02461_prewhere_row_level_policy_lightweight_delete` - https://github.com/ClickHouse/ClickHouse/pull/49764
- `02458_relax_too_many_parts`  - same

>Stress test (tsan) — Killed by signal (in clickhouse-server.log)

```
/var/log/clickhouse-server/clickhouse-server.err.log:2023.05.25 10:25:12.676037 [ 3405 ] {bb407f30-b95b-4e15-b8ea-3e1907538a7b} <Fatal> : Logical error: 'Expected 90639 >= 103598 (Buffer path: 00170_test/iky/veouleggbgovdglrwzdqvodwdqlpt, hash key: bb22b957e4bbead840959c0293fc1f03, file_offset_of_buffer_end: 103598, read_until_position: 2297058, internal buffer end: 103598, read_type: CACHED, last caller: bb407f30-b95b-4e15-b8ea-3e1907538a7b:3405, file segment info: File segment: [0, 172696], key: bb22b957e4bbead840959c0293fc1f03, state: PARTIALLY_DOWNLOADED, downloaded size: 90639, reserved size: 90639, downloader id: None, current write offset: 90639, first non-downloaded offset: 90639, caller id: bb407f30-b95b-4e15-b8ea-3e1907538a7b:3405, kind: Regular, unbound: 0)'.
```

- #50025

azat commented:
@devcrafter can you please take a look?

azat committed:
Adjust min_insert_block_size_rows for materialized_view_parallelize_output_from_storages

Otherwise it is too slow for perf tests on CI [1].

  [1]: https://s3.amazonaws.com/clickhouse-test-reports/50214/e287ec50920c7cadabea6ec19ef14b353345ac93/performance_comparison_[3_4]/report.html

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

devcrafter commented:
What is `input` in your example for materialized view?
```
INSERT INTO input SELECT * FROM numbers(10e6) SETTINGS parallelize_output_from_storages=0, min_insert_block_size_rows=1000
  0 rows in set. Elapsed: 1.851 sec. Processed 20.00 million rows, 120.00 MB (10.80 million rows/s., 64.82 MB/s.)
...
````

azat commented on src/Storages/System/StorageSystemZooKeeper.h:
```diff
@@ -35,6 +35,7 @@ class StorageSystemZooKeeper final : public IStorage
     bool isSystemStorage() const override { return true; }
     bool supportsIndexForIn() const override { return true; }
     bool mayBenefitFromIndexForIn(const ASTPtr & node, ContextPtr, const StorageMetadataPtr &) const override;
+    bool parallelizeOutputAfterReading(ContextPtr) const override { return false; }
```
I thought that maybe it could be enabled for `numbers_mt`, could be useful for tests for instance, but I'm ok with disabling it for all system tables.

devcrafter replied:
> I thought that maybe it could be enabled for numbers_mt

`numbers_mt` is not affected by this setting, since the pipeline is already paralleled by `number_mt` implementation. So, no resize is added [here](https://github.com/ClickHouse/ClickHouse/blob/22670aebf8fd54d2fc6f5c7cf64b8477729fcd02/src/Storages/IStorage.cpp#LL137-L141)

azat committed:
Disable parallelize_output_from_storages for all system tables

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat commented:
```
+ ln -s /usr/share/clickhouse-test/clickhouse-test /usr/bin/clickhouse-test
+ source /usr/share/clickhouse-test/ci/attach_gdb.lib
/run.sh: line 19: /usr/share/clickhouse-test/ci/attach_gdb.lib: No such file or directory
```

devcrafter committed:
Merge branch 'master' into parallelize_output_from_storages-fix

azat commented:
Test failures looks unrelated:

>Integration tests (release) [2/4] — fail: 1, passed: 560, flaky: 0

- `test_multiple_disks/test.py::test_background_move[mt]` - flaky

>Stateless tests (debug) [2/5] — fail: 2, passed: 1031, skipped: 10

- `02703_max_local_write_bandwidth` - #50935
- `02561_null_as_default_more_formats` - #49177

>Stateless tests (debug) [3/5] — fail: 1, passed: 1100, skipped: 18

- `01443_merge_truncate_long` - #50294

>Stateless tests (tsan, s3 storage) [3/5] — fail: 1, passed: 1084, skipped: 34

- `02481_async_insert_race_long` - #47866
